fms2_io/readme.md

```diff
@@ -0,0 +1,603 @@
+
+if (open_file(fileobj, "filename", "overwrite", pelist=pes, is_restart=.true.)) then
+  call register_axis(fileobj, dim_names(1), 96)
+  call register_axis(fileobj, dim_names(2), 96)
+  call register_axis(fileobj, dim_names(3), dimsize)
+  call register_axis(fileobj, dim_names(4), unlimited)
+
+  call register_field(fileobj, 'variable_name', 'variable_type', dim_names)
+  call write_data(fileobj)
+  call close_file(fileobj)
+endif
+```
+- `open_file` - the pelist argument is needed so that only the root pe writes the file
+- `register_axis` is needed so that the code knows the size of each dimension
+- `register_field` is needed so that the code knows what dimensions the variable is a function of
+- [write_data](https://github.com/NOAA-GFDL/FMS/blob/main/fms2_io/include/netcdf_read_data.inc): the root pe in the pelist writes the data
+
+To read non-domain decomposed restarts:
+
+```F90
+if (open_file(fileobj, "filename", "read", pelist=pes, is_restart=.true.)) then
+  call read_data(fileobj, 'variable_name', variable_data)
+  call close_file(fileobj)
+endif
+```
+- Here the `register_axis` and `register_field` calls are not needed because the code can get the dimensions from the file
+
+Additionally, the `corner` and `edge_lengths` optional arguments in the `read_data` calls can be used to read only a section of a file.
+```F90
+if (open_file(fileobj, "filename", "read", pelist=pes, is_restart=.true.)) then
+  call read_data(fileobj, 'variable_name', variable_data, corner=(/10,2,2,3/), edge_lengths=(/2, 3, 4, 1/))
+  call close_file(fileobj)
+endif
+```
+Here the code is going to start reading at x=10, y=2, z=2, t=3 and read 2, 3, 4, 1 points in each direction. Note that `variable_data` must be the correct size or the code will fail.
+
+Additionally, it is possible to do "compressed" reads and writes. Here "compressed" is when different ranks have different numbers of points for a specific
+axis. For example:
+```F90
+if (open_file(fileobj, "filename", "overwrite", pelist=pes, is_restart=.true.)) then
+  call register_axis(fileobj, "xaxis_1", mpp_pe()+1, is_compressed=.true.)
+  call register_axis(fileobj, "yaxis_1", mpp_pe()+1, is_compressed=.true.)
+
+  call register_field(fileobj, 'variable_name', 'variable_type', dim_names)
+  call write_data(fileobj)
+  call close_file(fileobj)
+endif
+```
+- Because [is_compressed=.true.](https://github.com/NOAA-GFDL/FMS/blob/b9fc6515c7e729909e59a0f9a1efc6eb1d3e44d1/fms2_io/netcdf_io.F90#L747-L768), the root pe is going to gather the dimension size of each rank and add them to get the total length of the dimension. In the example above, rank 0 has a x/y axis of 1, rank 1 has a x/y of 2. The total dimension size of x/y is equal to 3 (the sum for all pes).
+
+### E. Coupler Type Restarts
+With the functionality in `coupler_type_register_restarts`, one can loop through a list of variables in the coupler_type and register each variable to the corresponding restart file and then read/write the file.
```

Comment:
~~With the functionality in~~ `coupler_type_register_restarts`, ~~one can loop~~ permits iterating through (side note: after investigating how to phrase this, it turns out that "through", or any preposition, [is unnecessary](https://stackoverflow.com/questions/2715388/is-it-iterate-through-or-iterate-over-something) when using "iterating") list of variables in the coupler_type and register each variable to the corresponding restart file and then read/write the file.